### DISCUSSION TITLE
Using Singleton then create a new instance of HttpClient

### DISCUSSION THREAD

**Discussion by jakSolutionsGithub** [fuzzzerd/fmdata#188] [Q&A] [answered]
Hello,

Sorry for the possible dumb question.

I do not understand why in startup.cs we create a Singleton which contains the connection data to the FileMaker server and then in the examples, we must declare :

`FileMakerRestClient client = new FileMakerRestClient (serverName, fileName, userName, password);`

if we want to use FindAsyc() for exemple.
It creates a new instance of HttpClient, doesn't it?

Can you enlighten me?

**Comment by fuzzzerd** [comment, marked as answer]
The samples are intended to be a fully usable sample that is separate from using the asp.net dependency injection framework.

If you have a singleton defined in startup.cs, you should be able to just request an `IFileMakerApiClient` from the di container and use it directly.

**Reply by jakSolutionsGithub** [reply]
Hello again,

So I tried with the singleton as shown except when I get to the page I get the error:

`InvalidOperationException: Unable to resolve service for type 'FMData.Rest.IFileMakerRestClient' while attempting to activate 'cae.website.Controllers.Events.HistoryController'.`


This is how I created the singleton in startup.cs:

`            services.AddHttpClient(); // setup IHttpClientFactory in the DI container
            services.AddSingleton<FMData.ConnectionInfo>(ci => new FMData.ConnectionInfo
            {
                FmsUri = "https://myurl.test",
                Username = "testAccount",
                Password = "passwordAccount123",
                Database = "testDB"
            });
            // Keep the FileMaker client as a singleton for speed
            services.AddSingleton<IFileMakerApiClient, FileMakerRestClient>(s => {
                var hcf = s.GetRequiredService<IHttpClientFactory>();
                var ci = s.GetRequiredService<FMData.ConnectionInfo>();
                return new FileMakerRestClient(hcf.CreateClient(), ci);
            });`

Then in my Controller class I've declared a IFileMakerRestClient:

`private readonly IFileMakerRestClient client;`

And in my constructor, I did like this:

`        public ContactsController(IWebHostEnvironment env, IMapper mapper, IFileMakerRestClient client)
        {
            this.env = env;
            _mapper = mapper;
            this.client = client;
        }`

Am I doing something wrong here?

**Reply by fuzzzerd** [reply]
It looks like you are adding `IFileMakerApiClient` do your DI container, but in the controller you are requesting `IFileMakerRestClient`.
